Improving the handling of express method titles

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

This PR improves how we handle express payment method titles by introducing new constants and methods to replace duplicate code.

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

Code review should be enough. Check if tests are still passing.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Hiding actions for pending Amazon Pay orders

wjrosa committed:
Improving the handling of express method titles

wjrosa committed:
Checking for order received page

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Readme and changelog entries

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Revert automatic PHPCS fix

wjrosa committed:
Merge branch 'tweak/hiding-actions-for-pending-amazon-pay-orders' into dev/improving-handling-of-ece-method-titles

wjrosa commented on includes/payment-methods/class-wc-stripe-express-checkout-helper.php:
```diff
@@ -41,7 +41,37 @@ public function __construct() {
+	 */
+	public static function get_payment_method_titles( $include_link = false ) {
+		$titles = [
+			'apple_pay'                           => WC_Stripe_Express_Payment_Titles::APPLE_PAY,
```
Another implementation option would be translated names, like:
```php
'apple_pay' => __( 'Apple Pay', 'woocommerce-gateway-stripe' ),
```
But not sure if it is worth since those are brands, which I think are immutable.

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

diegocurbelo committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'dev/improving-handling-of-ece-method-titles' of https://github.com/woocommerce/woocommerce-gateway-stripe into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Fix changelog and readme entries

wjrosa committed:
Improving handling of BACS debit payment method title

Mayisha reviewed (commented):
Hi @wjrosa 

Thanks for the effort in this PR. After reviewing this, I think that the changes might not add significant value in improving the codebase. It is a bit overengineering to have a new class for the express method titles and also reduces the readability of the code. 
We have implemented constants for the payment methods in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3449. I think that was a good refactor and we should use that constant class to store the express method titles as well instead of making it complex with more new classes.

It would be better to hold off on merging this one. Let me know what you think and if you have a strong opinion in favor of these changes 👀

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Refactor implementation removing constant class following review comments

wjrosa committed:
Minor refactoring

wjrosa committed:
Fix tests + updating missing occurrences

wjrosa commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -515,14 +515,22 @@ public function add_stripe_intents( PaymentContext $context, PaymentResult &$res
-			$order->save();
-		} elseif ( 'payment_request_api' === $payment_request_type ) {
-			$order->set_payment_method_title( 'Payment Request (Stripe)' );
+		$payment_method_title = '';
```
Just making it easier to read. And fixing the suffix part (it is dynamic).

wjrosa committed:
Minor refactor

wjrosa commented on includes/payment-methods/class-wc-stripe-upe-payment-method-cc.php:
```diff
@@ -41,7 +41,11 @@ public function __construct() {
 	 */
 	public function get_title( $payment_details = false ) {
-		$wallet_type = WC_Stripe_Payment_Methods::AMAZON_PAY === ( $payment_details->type ?? null ) ? WC_Stripe_Payment_Methods::AMAZON_PAY : ( $payment_details->card->wallet->type ?? null );
+		if ( WC_Stripe_Payment_Methods::AMAZON_PAY === ( $payment_details->type ?? null ) ) {
```
Just making it easier to read.

wjrosa commented:
> Hi @wjrosa
> 
> Thanks for the effort in this PR. After reviewing this, I think that the changes might not add significant value in improving the codebase. It is a bit overengineering to have a new class for the express method titles and also reduces the readability of the code. We have implemented constants for the payment methods in #3449. I think that was a good refactor and we should use that constant class to store the express method titles as well instead of making it complex with more new classes.
> 
> It would be better to hold off on merging this one. Let me know what you think and if you have a strong opinion in favor of these changes 👀

Thanks, Mayisha! Oh OK, I have refactored it now just to follow the current implementation of the payment method labels (using the existing constant instead of adding the new one). I have also fixed all the title suffix calls (which were not checking for the filter value). Let me know what you think of this simpler approach 👀

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'develop' into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Merge branch 'dev/improving-handling-of-ece-method-titles' of https://github.com/woocommerce/woocommerce-gateway-stripe into dev/improving-handling-of-ece-method-titles

wjrosa committed:
Fix bad merge

malithsen commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -515,14 +515,22 @@ public function add_stripe_intents( PaymentContext $context, PaymentResult &$res
+				$payment_method_title = WC_Stripe_Payment_Methods::GOOGLE_PAY_LABEL;
+				break;
+			case 'payment_request_api':
+				$payment_method_title = 'Payment Request';
```
Should we move "Payment Request" to a constant as well?

wjrosa replied:
Sure! Did it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3844/commits/2454e16de799a32e341aefc00d1568200c97df4e

malithsen commented on tests/phpunit/payment-methods/test-wc-stripe-express-checkout-helper.php:
```diff
@@ -311,4 +311,16 @@ public function provide_test_get_normalized_postal_code() {
+	 * @return void
+	 */
+	public function test_get_payment_method_title_suffix() {
+		$wc_stripe_ece_helper = new WC_Stripe_Express_Checkout_Helper();
+		$actual               = $wc_stripe_ece_helper->get_payment_method_title_suffix();
```
Since `get_payment_method_title_suffix` is a static method, instantiating the class is unnecessary. 

```suggestion
		$actual = WC_Stripe_Express_Checkout_Helper::get_payment_method_title_suffix();
```

malithsen commented on includes/payment-methods/class-wc-stripe-express-checkout-helper.php:
```diff
@@ -41,7 +41,19 @@ public function __construct() {
+	/**
+	 * Returns the suffix set for the express payment method titles.
+	 *
+	 * @return mixed
```
Shouldn't this method always return a string?

```suggestion
	 * @return string
```

malithsen reviewed (approved):
I've reviewed the changes made since @Mayisha's feedback.

The implementation is now much cleaner after removing the separate constant class for method titles.

I'm approving this PR but I've nitpicked a couple of minor things in comments. Would be great if you want to address them before the merge but okay with merging as-is.

wjrosa committed:
Update includes/payment-methods/class-wc-stripe-express-checkout-helper.php

Co-authored-by: Malith Senaweera <6216000+malithsen@users.noreply.github.com>

wjrosa committed:
Update tests/phpunit/payment-methods/test-wc-stripe-express-checkout-helper.php

Co-authored-by: Malith Senaweera <6216000+malithsen@users.noreply.github.com>

wjrosa committed:
Including a payment label for payment request
